python/test/unit/mesh/test_mesh.py

```diff
@@ -205,6 +205,102 @@ def test_mesh_construction_pygmsh():
     assert mesh.topology.dim == 2
 
 
+def test_mesh_construction_gmsh():
+    import gmsh
+
+    # Map from gmsh to type that is also used by meshio
+    mapping = {1: 'line',
+               2: 'triangle',
+               4: 'tetra',
+               8: 'line3',
+               9: 'triangle6',
+               11: 'tetra10',
+               15: 'vertex'
+               }
+    nodes = {'line': 2,
+             'triangle': 3,
+             'tetra': 4,
+             'line3': 3,
+             'triangle6': 6,
+             'tetra10': 10,
+             'vertex': 1
+             }
+    # Generate mesh
+    gmsh.initialize(sys.argv)
+    gmsh.model.add('ball')
+    gmsh.model.occ.addSphere(0.0, 0.0, 0.0, 1.0)
+    gmsh.model.occ.synchronize()
+    gmsh.model.mesh.generate()
+    nodeTags, coord, parametricCoord = gmsh.model.mesh.getNodes()
+    dim = 3
+    # Reshape to get right format
+    points = numpy.reshape(coord, (numpy.int(coord.size / dim), dim))
+    elementTypes, elementTags, nodeTags = gmsh.model.mesh.getElements()
+    cells = {}
+    # Generate cells dict
+    for num, element in enumerate(elementTypes):
+        name = mapping[element]
+        dim = nodes[name]
+        cells[name] = nodeTags[num] - 1  # since nodes are numbered starting from 0
+        cells[name] = numpy.reshape(cells[name], (numpy.int(cells[name].size / dim), dim))
+    # End Generate mesh
+    gmsh.finalize()
+
+    mesh = dolfin.cpp.mesh.Mesh(dolfin.MPI.comm_world, dolfin.cpp.mesh.CellType.Type.tetrahedron, points,
+                                cells['tetra'], [], dolfin.cpp.mesh.GhostMode.none)
+    assert mesh.degree() == 1
+    assert mesh.geometry.dim == 3
+    assert mesh.topology.dim == 3
+
+    mesh = dolfin.cpp.mesh.Mesh(dolfin.MPI.comm_world,
+                                dolfin.cpp.mesh.CellType.Type.triangle, points,
+                                cells['triangle'], [], dolfin.cpp.mesh.GhostMode.none)
+    assert mesh.degree() == 1
+    assert mesh.geometry.dim == 3
+    assert mesh.topology.dim == 2
+
+    mesh = dolfin.cpp.mesh.Mesh(dolfin.MPI.comm_world,
+                                dolfin.cpp.mesh.CellType.Type.interval, points,
+                                cells['line'], [], dolfin.cpp.mesh.GhostMode.none)
+    assert mesh.degree() == 1
+    assert mesh.geometry.dim == 3
+    assert mesh.topology.dim == 1
+
+    gmsh.initialize(sys.argv)
+    # Generate 2nd order mesh
+    gmsh.model.add('ball')
+    gmsh.model.occ.addSphere(0.0, 0.0, 0.0, 1.0)
+    gmsh.model.occ.synchronize()
+    gmsh.model.mesh.generate()
+    gmsh.model.mesh.setOrder(2)
+    nodeTags, coord, parametricCoord = gmsh.model.mesh.getNodes()
+    dim = 3
+    # Reshape to get right format
+    points = numpy.reshape(coord, (numpy.int(coord.size / dim), dim))
```

Comment:
Could code form the first part of the test be re-used? Maybe use pytest  test parameterisation.